File: launcher-tool/src/main/kotlin/IO.kt
package dk.sdu.cloud

sealed class FileFactory {
    abstract fun create(path: String): LFile
}

sealed class LFile(val path: String) {
    abstract val absolutePath: String
    val name: String = path.substringAfterLast('/')

    abstract fun exists(): Boolean
    abstract fun child(subpath: String): LFile
    abstract fun writeText(text: String)
    abstract fun writeBytes(bytes: ByteArray)
    abstract fun appendText(text: String)
    abstract fun delete()
    abstract fun mkdirs()

    override fun toString() = absolutePath
}

sealed class ExecutableCommandFactory {
    abstract fun create(
        args: List<String>,
        workingDir: LFile? = null,
        postProcessor: (result: ProcessResultText) -> String = { it.stdout },
        allowFailure: Boolean = false,
        deadlineInMillis: Long = 1000 * 60 * 5,
        streamOutput: Boolean = false,
    ): ExecutableCommand
}

interface ExecutableCommand {
    val args: List<String>
    val workingDir: LFile?
    val postProcessor: (result: ProcessResultText) -> String
    var allowFailure: Boolean
    var deadlineInMillis: Long
    var streamOutput: Boolean

    fun toBashScript(): String
    fun executeToText(): Pair<String?, String>
}

fun <T : ExecutableCommand> T.streamOutput(): T {
    streamOutput = true
    return this
}

fun <T : ExecutableCommand> T.allowFailure(): T {
    allowFailure = true
    return this
}

data class ProcessResultText(
    val statusCode: Int,
    val stdout: String,
    val stderr: String,
)

File: launcher-tool-go/pkg/launcher/io.go
package launcher

import (
	"strings"
)

type LFile interface {
	GetAbsolutePath() string
	Exists() bool
	Child(subPath string, isDir bool) LFile
	WriteText(text string)
	WriteBytes(bytes []byte)
	AppendText(text string)
	Delete()
	MkDirs()
	Name() string
}

type ExecutableCommandInterface interface {
	ToBashScript() string
	ExecuteToText() StringPair
	SetStreamOutput()
	SetAllowFailure()
	SetDeadline(DeadlineInMillis int64)
}

func NewFile(path string) LFile {
	return LocalFile{
		path: path,
	}
}

func NewExecutableCommand(
	args []string,
	workingDir LFile,
	fn postProcessor,
) ExecutableCommandInterface {
	return NewLocalExecutableCommand(
		args,
		workingDir,
		fn,
	)
}

type postProcessor func(text ProcessResultText) string

func PostProcessorFunc(text ProcessResultText) string {
	return text.stdout
}

type ProcessResultText struct {
	statusCode int
	stdout     string
	stderr     string
}

// NOTE(Dan): This is slightly different from how escapeBash works in the Kotlin version since this also automatically
// wraps it in single quotes. This is how it was used in all cases anyway, so this makes it slightly simpler to use.

func EscapeBash(s string) string {
	builder := &strings.Builder{}
	builder.WriteRune('\'')
	for _, c := range []rune(s) {
		if c == '\'' {
			builder.WriteString("'\"'\"'")
		} else {
			builder.WriteRune(c)
		}
	}
	builder.WriteRune('\'')
	return builder.String()
}
